<a href="https://colab.research.google.com/github/kibarmustofa/2341720034_ML_2025/blob/main/JS03/TG1_JS03_2341720034_KIBAR_MUSTOFA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Tugas Praktikum
Wisconsin Breast Cancer (again)

nama : Kibar Mustofa

kelas: TI3F

absen: 17

#Deskripsi Tugas
#
Pada tugas pratikum ini Anda akan menggunakan data "Wisconsin Breast Cancer". Data tersebut terdiri dari 569 data yang digunakan untuk mendiagnonis jenis kanker Malignant (M) dan Benign (B). Tugas Anda adalah,

link : /content/drive/MyDrive/wbc.csv

1. Pisahkan antara variabel yang dapat digunakan dan variabel yang tidak dapat digunakan.

2. Lakukan proses encoding pada kolom "diagnosis".

3. Lakukan proses standardisasi pada semua kolom yang memiliki nilai numerik.

4. Lakukan proses seleksi fitur. Anda dapat menggunakan SelectKBest.

5. Lakukan proses pengujian dengan model Logistic Regression seperti pada praktikum 1.

6. Anda dapat menggunakan model pipeline untuk mempermudah perkejaan Anda.

7. Berdasarkan hasil analisa Anda, berapa jumlah fitur terbaik yang dapat digunakan? Apa saja fitur tersebut?

#Memuat data

In [ ]:
import pandas as pd
import numpy as np
from sklearn.datasets import load_breast_cancer
from sklearn.preprocessing import LabelEncoder, StandardScaler, MinMaxScaler
from sklearn.feature_selection import SelectKBest, f_classif
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import Pipeline
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import accuracy_score, classification_report

In [ ]:
df = pd.read_csv('/content/drive/MyDrive/wbc.csv')
df.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0  ...          17.33           184.60      2019.0            0.1622   
1  ...          23.41           158.80      1956.0            0.1238   
2  ...          25.53           152.50      1709.0            0.1444   
3  ...          26.50            98.87       567.7            0.2098   
4  ...          16.67           152.20      1575.0            0.1374   

   compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN  

[5 rows x 33 columns]

#1. Memisahkan variabel
Pisahkan fitur (variabel yang dapat digunakan) dari target (variabel yang tidak dapat digunakan untuk prediksi, dalam kasus ini 'diagnosis').

In [ ]:
# Separate features (X) and target (y)
x = df.drop(['id', 'Unnamed: 32'], axis=1)
y = df['diagnosis']

display(x.head())
display(y.head())

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0         M        17.99         10.38          122.80     1001.0   
1         M        20.57         17.77          132.90     1326.0   
2         M        19.69         21.25          130.00     1203.0   
3         M        11.42         20.38           77.58      386.1   
4         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0         0.2419  ...         25.38          17.33           184.60   
1         0.1812  ...         24.99          23.41           158.80   
2         0.2069  ...         23.57          25.53           152.50   
3         0.2597  ...         14.91          26.50            98.87   
4         0.1809  ...         22.54          16.67           152.20   

   area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   

   concave points_worst  symmetry_worst  fractal_dimension_worst  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  
3                0.2575          0.6638                  0.17300  
4                0.1625          0.2364                  0.07678  

[5 rows x 31 columns]

0    M
1    M
2    M
3    M
4    M
Name: diagnosis, dtype: object

#2. Encoding Target
Lakukan encoding pada kolom 'diagnosis' menggunakan LabelEncoder.

In [ ]:
# Encode the target variable
le = LabelEncoder()
x['diagnosis'] = le.fit_transform(x['diagnosis'])
x.head(10)

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0          1        17.99         10.38          122.80     1001.0   
1          1        20.57         17.77          132.90     1326.0   
2          1        19.69         21.25          130.00     1203.0   
3          1        11.42         20.38           77.58      386.1   
4          1        20.29         14.34          135.10     1297.0   
5          1        12.45         15.70           82.57      477.1   
6          1        18.25         19.98          119.60     1040.0   
7          1        13.71         20.83           90.20      577.9   
8          1        13.00         21.82           87.50      519.8   
9          1        12.46         24.04           83.97      475.9   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760         0.30010              0.14710   
1          0.08474           0.07864         0.08690              0.07017   
2          0.10960           0.15990         0.19740              0.12790   
3          0.14250           0.28390         0.24140              0.10520   
4          0.10030           0.13280         0.19800              0.10430   
5          0.12780           0.17000         0.15780              0.08089   
6          0.09463           0.10900         0.11270              0.07400   
7          0.11890           0.16450         0.09366              0.05985   
8          0.12730           0.19320         0.18590              0.09353   
9          0.11860           0.23960         0.22730              0.08543   

   symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0         0.2419  ...         25.38          17.33           184.60   
1         0.1812  ...         24.99          23.41           158.80   
2         0.2069  ...         23.57          25.53           152.50   
3         0.2597  ...         14.91          26.50            98.87   
4         0.1809  ...         22.54          16.67           152.20   
5         0.2087  ...         15.47          23.75           103.40   
6         0.1794  ...         22.88          27.66           153.20   
7         0.2196  ...         17.06          28.14           110.60   
8         0.2350  ...         15.49          30.73           106.20   
9         0.2030  ...         15.09          40.68            97.65   

   area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   
5       741.6            0.1791             0.5249           0.5355   
6      1606.0            0.1442             0.2576           0.3784   
7       897.0            0.1654             0.3682           0.2678   
8       739.3            0.1703             0.5401           0.5390   
9       711.4            0.1853             1.0580           1.1050   

   concave points_worst  symmetry_worst  fractal_dimension_worst  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  
3                0.2575          0.6638                  0.17300  
4                0.1625          0.2364                  0.07678  
5                0.1741          0.3985                  0.12440  
6                0.1932          0.3063                  0.08368  
7                0.1556          0.3196                  0.11510  
8                0.2060          0.4378                  0.10720  
9                0.2210          0.4366                  0.20750  

[10 rows x 31 columns]

#3. Standarisasi pada kolom numerik
Lakukan proses standardisasi pada semua kolom yang memiliki nilai numerik.

In [ ]:
std = StandardScaler()
# Select numerical columns from X
numerical_cols = x.columns
x[numerical_cols] = std.fit_transform(x[numerical_cols])
display(x.head(10))

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0   1.297676     1.097064     -2.073335        1.269934   0.984375   
1   1.297676     1.829821     -0.353632        1.685955   1.908708   
2   1.297676     1.579888      0.456187        1.566503   1.558884   
3   1.297676    -0.768909      0.253732       -0.592687  -0.764464   
4   1.297676     1.750297     -1.151816        1.776573   1.826229   
5   1.297676    -0.476375     -0.835335       -0.387148  -0.505650   
6   1.297676     1.170908      0.160649        1.138125   1.095295   
7   1.297676    -0.118517      0.358450       -0.072867  -0.218965   
8   1.297676    -0.320167      0.588830       -0.184080  -0.384207   
9   1.297676    -0.473535      1.105439       -0.329482  -0.509063   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0         1.568466          3.283515        2.652874             2.532475   
1        -0.826962         -0.487072       -0.023846             0.548144   
2         0.942210          1.052926        1.363478             2.037231   
3         3.283553          3.402909        1.915897             1.451707   
4         0.280372          0.539340        1.371011             1.428493   
5         2.237421          1.244335        0.866302             0.824656   
6        -0.123136          0.088295        0.300072             0.646935   
7         1.604049          1.140102        0.061026             0.281950   
8         2.201839          1.684010        1.219096             1.150692   
9         1.582699          2.563358        1.738872             0.941760   

   symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0       2.217515  ...      1.886690      -1.359293         2.303601   
1       0.001392  ...      1.805927      -0.369203         1.535126   
2       0.939685  ...      1.511870      -0.023974         1.347475   
3       2.867383  ...     -0.281464       0.133984        -0.249939   
4      -0.009560  ...      1.298575      -1.466770         1.338539   
5       1.005402  ...     -0.165498      -0.313836        -0.115009   
6      -0.064325  ...      1.368983       0.322883         1.368325   
7       1.403355  ...      0.163763       0.401048         0.099449   
8       1.965600  ...     -0.161357       0.822813        -0.031609   
9       0.797298  ...     -0.244190       2.443109        -0.286278   

   area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0    2.001237          1.307686           2.616665         2.109526   
1    1.890489         -0.375612          -0.430444        -0.146749   
2    1.456285          0.527407           1.082932         0.854974   
3   -0.550021          3.394275           3.893397         1.989588   
4    1.220724          0.220556          -0.313395         0.613179   
5   -0.244320          2.048513           1.721616         1.263243   
6    1.275220          0.518640           0.021215         0.509552   
7    0.028859          1.447961           0.724786        -0.021054   
8   -0.248363          1.662757           1.818310         1.280035   
9   -0.297409          2.320295           5.112877         3.995433   

   concave points_worst  symmetry_worst  fractal_dimension_worst  
0              2.296076        2.750622                 1.937015  
1              1.087084       -0.243890                 0.281190  
2              1.955000        1.152255                 0.201391  
3              2.175786        6.046041                 4.935010  
4              0.729259       -0.868353                -0.397100  
5              0.905888        1.754069                 2.241802  
6              1.196716        0.262476                -0.014730  
7              0.624196        0.477640                 1.726435  
8              1.391616        2.389857                 1.288650  
9              1.620015        2.370444                 6.846856  

[10 rows x 31 columns]

# 4. Lakukan proses seleksi fitur. Anda dapat menggunakan SelectKBest.

In [ ]:
#selector = SelectKBest(f_classif, k=10) namun saya menggunakan pipeline

#5.Lakukan proses pengujian dengan model Logistic Regression seperti pada praktikum 1.

In [ ]:
#clf = LogisticRegression()

# 6. Anda dapat menggunakan model pipeline untuk mempermudah perkejaan Anda.


In [ ]:
pipe_filter = Pipeline([
    ("std", StandardScaler()),
    ("selector", SelectKBest(f_classif, k= 20)),
    ("clf", LogisticRegression(max_iter=1000))
])

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, test_size=0.2, random_state=42)
pipe_filter.fit(X_train, y_train)
pred = pipe_filter.predict(X_test)
print("=== Filter (ANOVA) + LR ===")
print("Accuracy:", accuracy_score(y_test, pred))
print(classification_report(y_test, pred))

=== Filter (ANOVA) + LR ===
Accuracy: 0.9912280701754386
              precision    recall  f1-score   support

           B       0.99      1.00      0.99        72
           M       1.00      0.98      0.99        42

    accuracy                           0.99       114
   macro avg       0.99      0.99      0.99       114
weighted avg       0.99      0.99      0.99       114



In [ ]:
# 1) Nama fitur setelah preprocess
feat_names = pipe_filter.named_steps["std"].get_feature_names_out()
print("Nama fitur:", feat_names)
print("\n")

# 2) Mask & skor fitur terpilih (SelectKBest)
sel = pipe_filter.named_steps["selector"]
mask = sel.get_support()
selected_names = feat_names[mask]
selected_scores = sel.scores_[mask]
top = sorted(zip(selected_names, selected_scores), key=lambda t: t[1], reverse=True)[:10]
print("Top fitur:", top)

Nama fitur: ['radius_mean' 'texture_mean' 'perimeter_mean' 'area_mean'
 'smoothness_mean' 'compactness_mean' 'concavity_mean'
 'concave points_mean' 'symmetry_mean' 'fractal_dimension_mean'
 'radius_se' 'texture_se' 'perimeter_se' 'area_se' 'smoothness_se'
 'compactness_se' 'concavity_se' 'concave points_se' 'symmetry_se'
 'fractal_dimension_se' 'radius_worst' 'texture_worst' 'perimeter_worst'
 'area_worst' 'smoothness_worst' 'compactness_worst' 'concavity_worst'
 'concave points_worst' 'symmetry_worst' 'fractal_dimension_worst'
 'area_perimeter_ratio']


Top fitur: [('concave points_worst', np.float64(733.7249325739482)), ('perimeter_worst', np.float64(717.2464871041376)), ('radius_worst', np.float64(692.861395260564)), ('concave points_mean', np.float64(684.5268452011383)), ('perimeter_mean', np.float64(548.4132358502825)), ('area_worst', np.float64(522.1889466673214)), ('radius_mean', np.float64(511.27484784479856)), ('area_perimeter_ratio', np.float64(470.77078841261965)), ('area_m

#7. Berdasarkan hasil analisa Anda, berapa jumlah fitur terbaik yang dapat digunakan? Apa saja fitur tersebut?

- percobaan ini saya menggunakan SelectKBest dengan k=20, modelmu mencapai akurasi 0.9912. Sepuluh fitur teratas dengan skor tertinggi adalah:

1. concave points_worst
2. perimeter_worst
3. radius_worst
4. concave points_mean
5. perimeter_mean
6. area_worst
7. radius_mean
8. area_perimeter_ratio
9. area_mean
10. concavity_mean
dst.

tapi, melihat hasil akurasi yang sudah tinggi dengan 20 fitur ini, sepertinya fitur-fitur tersebut memang sangat penting untuk memprediksi jenis kanker.